### PR TITLE
Upgrade go-parser

### PR DESCRIPTION
**Pull request by joewagner**
Upgrade the sql parser to use the latest version of go-sqlparser.
This must be released after the Validator has been deployed with the same version.

related to:
https://github.com/tablelandnetwork/go-sqlparser/pull/52
https://github.com/tablelandnetwork/go-tableland/pull/542

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -2,4 +2,7 @@ module github.com/tablelandnetwork/wasm-sqlparser
 
 go 1.16
 
-require github.com/tablelandnetwork/sqlparser v0.0.0-20230328132500-785ebca8e351
+require (
+	github.com/hashicorp/errwrap v1.1.0 // indirect
```
Comment (joewagner) [inline go.mod:6]: I did `go get -u github.com/tablelandnetwork/sqlparser` and it included this line.  Anyone know why?

**Comment by joewagner** [comment]
It seems like the ens table name mapping tests are failing after the update to `go-sqlparser` made in https://github.com/tablelandnetwork/go-sqlparser/pull/52.  Maybe something to do with how back ticks are being handled.  
cc: @asutula

File: go.mod (hunk 1 above)
Comment (asutula) [reply go.mod:6]: Not sure. Run `go mod tidy` just to make sure things are fine.

**Comment by asutula** [comment]
Hey @joewagner i think the failing tests are because the identifier quotes, in this case back ticks, are now preserved. This is a more correct behavior that we wanted. So the tests just need to be updated I believe.

**Comment by joewagner** [comment]
I'm looking at this again, and the tests that are failing have to do with mapping table names. 
for example:
When you do, 
```
await globalThis.sqlparser.normalize(
    "select `t1`.id, t3.* from t1, t2 join t3 join (select * from t4);",
    { t1: "table1", t2: "table2", t3: "table3" } // Leave t4 "as is"
);
```
The resulting statement is:
```select `t1`.id,table3.* from table1 join table2 join table3 join(select * from t4)```.
Specifically the "t1" inside the backticks isn't being mapped to "table1".   Is that the correct behaviour? 

I don't have much context here, but I think we would want to do the mapping and preserve backticks. 
For example, the output would look like this:
```select `table1`.id,table3.* from table1 join table2 join table3 join(select * from t4)```

The CLI relies on the name mapping functionality for the ENS feature flag. The only tests in the CLI for the ENS feature are in [an in progress PR](https://github.com/tablelandnetwork/js-tableland-cli/pull/272/files) so it's pretty hard to tell what the correct behavior is.

cc: @carsonfarmer @awmuncy

**Comment by carsonfarmer** [comment]
Wait, are you sure that's what we want @asutula? That doesn't seem right to me? We also have some tests in js-tableland that pull in a 3rd party ORM that relies on some of the backtick behavior, but I'm not sure if this change would break that or not?

**Comment by carsonfarmer** [comment]
I guess it is possible this is the issue: https://github.com/tablelandnetwork/wasm-sqlparser/blob/main/main.go#L14. Or is it that the name from the AST is coming back _with_ the backticks along for the ride?

**Comment by carsonfarmer** [comment]
Right, ok this makes sense. I think solution #2 makes the most sense to me. In the short term, we can for sure deal with this in wasm-sqlparser by updating the regex to allow for wrapping the identifier in escape quotes, backticks, etc. It should actually be a pretty simple update, because we know it will have at most one of [`"'] at the start, and at most one of [`"'] at the end.

**Comment by asutula** [comment]
Hey guys, glad we're making progress here. Sounds promising. Just want to re-enforce a general rule I think applies here: If the author of a query is quoting any part of their query string in a supported way (using ", `, [, whatever), those quote characters should be seen by the the validator's sqlite db query parser/executor as is. If we must convert one type of quotes to some canonical type, I guess that's fine, but I'm afraid that is maybe just another assumption we make that we don't fully understand the consequences of.

**Comment by carsonfarmer** [comment]
Yes, 100% agree that these quotes need to be passing along, and should not be altered. I would say we should try to avoid canonicalizing the particular escape quotes as well. If you pass in ``, it should make it all the way to the SQL engine. But, we do need to be able to treat that thing within the quotes as an identifier, as the AST understands it. For replacements, and also to ensure that strings are clearly separate things from identifiers.

File: main.go [hunk 2]
```diff
@@ -30,9 +44,15 @@ func UpdateTableNames(node sqlparser.Node, nameMapper func(string) (string, bool
...
+
+				tableName, enclosure, isEnclosed := getEnclosedName(tableName)
+
 				if !tableNameRegEx.MatchString(tableName) {
 					return true, &sqlparser.ErrTableNameWrongFormat{Name: tableName}
 				}
+				if isEnclosed {
+					tableName = enclosure.open + tableName + enclosure.close
+				}
 				table.Name = sqlparser.Identifier(tableName)
```
Comment (joewagner) [inline main.go:56]: taking the name out of the enclosure to enable the regex to validate the name, then putting it back in the enclosure if appropriate.

File: main.go [hunk 3]
```diff
@@ -22,6 +23,19 @@ const (
...
+type EnclosingType struct {
+	open  string
+	close string
+}
+
+func getEnclosures() []EnclosingType {
+	return []EnclosingType{
+		{ open: "`", close: "`" },
+		{ open: "\"", close: "\"" },
+		{ open: "[", close: "]" },
```
Comment (joewagner) [inline main.go:35]: Any other types of enclosures we want to support?

File: main.go [hunk 4]
```diff
@@ -136,7 +156,20 @@ func normalize(this js.Value, args []js.Value) interface{} {
...
+					var value js.Value
+					_name, enclosure, isEnclosed := getEnclosedName(name)
+
+					if isEnclosed {
+						value = nameMap.Get(_name)
+						if value.IsUndefined() {
+							return "", false
+						}
+						return enclosure.open + value.String() + enclosure.close, true
+					}
```
Comment (joewagner) [inline main.go:170]: Take the name out of the enclosure and try to do the mapping and put the name back in.

File: main.go [hunk 5]
```diff
@@ -182,6 +215,21 @@ func normalize(this js.Value, args []js.Value) interface{} {
...
+	for _, encloseChar := range getEnclosures() {
+		if strings.HasPrefix(name, encloseChar.open) && strings.HasSuffix(name, encloseChar.close) {
+			_, _name, _ = strings.Cut(name, encloseChar.open)
+			_name, _, _ = strings.Cut(_name, encloseChar.close)
+			_enclosure = encloseChar
+			return _name, _enclosure, true
+		}
+	}
+
+	return name, EnclosingType{open: "", close: ""}, false
```
Comment (joewagner) [inline main.go:230]: This function takes a name node string and if it is enclosed returns the actual name, the enclosure characters, and a bool indicator.  If the name is not enclosed it returns the original name.

**Review by joewagner — COMMENTED** [review]
I've tried to implement the work around that will get this working until tactic <span>#</span>2 can be implemented in the go-parser.

File: test/main.test.js [hunk 6]
```diff
@@ -215,10 +215,11 @@ describe("sqlparser", function () {
...
         { t1: "table1", t2: "table2", t3: "table3" } // Leave t4 "as is"
       );
+
       // Note the canonical "join" added below to replace the comma
       strictEqual(
         statements.join(""),
-        "select table1.id,table3.* from table1 join table2 join table3 join(select * from t4)"
+        "select `table1`.id,table3.* from table1 join table2 join table3 join(select * from t4)"
       );
       deepStrictEqual(tables, ["table1", "table2", "table3", "t4"]);
```
Comment (joewagner) [inline test/main.test.js:224]: Now the return value from normalize leaves the escapes in the statement string, but the table names do not include the escape chars.

**Review by joewagner — COMMENTED** [review]
I made the changes we discussed today.  This is ready for another review.

**Review by asutula — APPROVED** [review]

**Pull request merged**